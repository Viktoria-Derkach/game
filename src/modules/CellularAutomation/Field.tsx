import React, { FC, useEffect, useRef } from 'react';

import { getColor } from './filling';
import { Cell } from './types';

export interface Props {
  field: Cell[][];
  cellSize?: number;
  width: number;
  height: number;
}

export const Field: FC<Props> = ({
  field,
  width,
  height,
  cellSize: cs = 4,
}) => {
  const canvasRef = useRef<HTMLCanvasElement>(null);

  useEffect(() => {
    const canvas = canvasRef.current;
    if (canvas) {
      const context = canvas.getContext('2d');
      if (context) {
        for (let y = 0; y < height; y++) {
          for (let x = 0; x < width; x++) {
            const state = field[y][x];
            context.fillStyle = getColor(state);
            context.fillRect(x * cs, y * cs, cs, cs);
          }
        }
      }
    }
  }, [field, width, height, cs]);

  return (
    <CanvasSection
      canvasRef={canvasRef}
      width={width * cs}
      height={height * cs}
    />
  );
};

const CanvasSection: FC<{
  canvasRef: React.RefObject<HTMLCanvasElement>;
  width: number;
  height: number;
}> = React.memo(({ canvasRef, width, height }) => {
  return (
    <section>
      <canvas ref={canvasRef} width={width} height={height}></canvas>
    </section>
  );
});

CanvasSection.displayName = 'CanvasSection';
