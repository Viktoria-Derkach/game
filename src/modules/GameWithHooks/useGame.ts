import { useState, useMemo, useEffect } from 'react';

import {
  Field,
  CellState,
  generateFieldWithDefaultState,
  fieldGenerator,
  Coords,
} from '@/helpers/Field';
import { openCell } from '@/helpers/openCell';
import { setFlag } from '@/helpers/setFlag';

import { LevelNames, GameSettings } from '@/modules/GameSettings';

interface ReturnType {
  level: LevelNames;
  time: number;
  isGameOver: boolean;
  isWin: boolean;
  settings: [number, number];
  playerField: Field;
  gameField: Field;
  onClick: (coords: Coords) => void;
  onContextMenu: (coords: Coords) => void;
  onChangeLevel: (level: LevelNames) => void;
  onReset: () => void;
}

export const useGame = (): ReturnType => {
  const [level, setLevel] = useState<LevelNames>('beginner');

  const [isGameOver, setIsGameOver] = useState(false);
  const [isWin, setIsWin] = useState(false);
  const [isGameStarted, setIsGameStarted] = useState(false);

  const [time, setTime] = useState(0);

  const setGameOver = (isSolved = false) => {
    setIsGameOver(true);
    setIsWin(isSolved);
  };

  const [size, bombs] = GameSettings[level];

  const [playerField, setPlayerField] = useState<Field>(
    generateFieldWithDefaultState(size, CellState.hidden)
  );

  const [gameField, setGameField] = useState<Field>(
    fieldGenerator(size, bombs / (size * size))
  );

  useEffect(() => {
    let interval: NodeJS.Timeout;

    if (isGameStarted) {
      interval = setTimeout(() => {
        setTime(time + 1);
      }, 1000);

      if (isGameOver) {
        clearInterval(interval);
      }
    }

    return () => {
      clearInterval(interval);
    };
  }, [isGameOver, isGameStarted, time]);

  useMemo(() => console.log(gameField), []);

  const onClick = (coords: Coords) => {
    !isGameStarted && setIsGameStarted(true);
    try {
      const [newPlayerField, isSolved, flagCounter] = openCell(
        coords,
        playerField,
        gameField
      );
      if (isSolved) {
        setGameOver(isSolved);
      }
      setPlayerField([...newPlayerField]);
    } catch (e) {
      setPlayerField([...gameField]);
      setGameOver(false);
    }
  };

  const onContextMenu = (coords: Coords) => {
    !isGameStarted && setIsGameStarted(true);
    const [newPlayerField, isSolved, flagCounter] = setFlag(
      coords,
      playerField,
      gameField
    );
    if (isSolved) {
      setGameOver(isSolved);
    }
    setPlayerField([...newPlayerField]);
  };

  const resetHandler = ([size, bombs]: [number, number]) => {
    const newGameField = fieldGenerator(size, bombs / (size * size));
    const newPlayerField = generateFieldWithDefaultState(
      size,
      CellState.hidden
    );

    setGameField([...newGameField]);
    setPlayerField([...newPlayerField]);
    setIsGameOver(false);
    setIsWin(false);
    setIsGameStarted(false);
    setTime(0);
  };

  const onChangeLevel = (level: LevelNames) => {
    setLevel(level);
    const newSettings = GameSettings[level];
    resetHandler(newSettings);
  };

  const onReset = () => resetHandler([size, bombs]);

  return {
    level,
    time,
    isGameOver,
    isWin,
    settings: [size, bombs],
    playerField,
    gameField,
    onClick,
    onContextMenu,
    onChangeLevel,
    onReset,
  };
};
