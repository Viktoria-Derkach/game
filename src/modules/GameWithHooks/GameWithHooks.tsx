import React, { FC, useCallback } from 'react';

import { GameLevels, LevelNames } from '@/modules/GameSettings';
import { Scoreboard } from '@/components/Scoreboard';
import { Grid } from '@/components/Grid';
import { GameOver } from '@/components/Game';
import { useGame } from './useGame';

export const GameWithHooks: FC = () => {
  const {
    level,
    isGameOver,
    isWin,
    settings,
    playerField,
    flagCounter,
    onClick,
    onContextMenu,
    onChangeLevel,
    onReset,
    time,
  } = useGame();

  const [, bombs] = settings;

  const onChangeLevelHandler = useCallback(
    ({ target: { value: level } }: React.ChangeEvent<HTMLSelectElement>) =>
      onChangeLevel(level as LevelNames),
    []
  );

  return (
    <>
      <Scoreboard
        time={String(time)}
        bombs={String(bombs - flagCounter)}
        levels={GameLevels as unknown as string[]}
        defaultLevel={level}
        onChangeLevel={onChangeLevelHandler}
        onReset={onReset}
      />
      {isGameOver && <GameOver onClick={onReset} isWin={isWin} />}
      <Grid onClick={onClick} onContextMenu={onContextMenu}>
        {playerField}
      </Grid>
    </>
  );
};
