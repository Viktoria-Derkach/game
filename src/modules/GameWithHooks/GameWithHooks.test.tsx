import React from 'react';
import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';

import { CellState } from '@/core/Field';

import { GameWithHooks } from './GameWithHooks';

import { useQuery } from '@/hooks/useQuery';

const mockOnClick = jest.fn();
const mockOnChangeLevel = jest.fn();
const mockOnReset = jest.fn();
const mockOnContextMenu = jest.fn();
const mockHistoryPush = jest.fn();

jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

jest.mock('@/hooks/useQuery', () => ({
  __esModule: true,
  useQuery: jest.fn(),
}));

(useQuery as jest.Mock).mockReturnValue({ get: () => null });

jest.mock('./useGame', () => ({
  __esModule: true,
  useGame: (level = 'beginner') => ({
    level,
    time: 0,
    flagCounter: 0,
    isGameOver: true,
    isWin: false,
    settings: [9, 10],
    playerField: [
      [10, 10],
      [10, 10],
    ],
    onClick: mockOnClick,
    onContextMenu: mockOnContextMenu,
    onChangeLevel: mockOnChangeLevel,
    onReset: mockOnReset,
  }),
}));

beforeEach(() => {
  jest.clearAllMocks();
});

describe('GameWithHooks test cases', () => {
  it('Render game field by default', () => {
    const { asFragment } = render(<GameWithHooks />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('Cell click works fine', () => {
    render(<GameWithHooks />);
    userEvent.click(screen.getByTestId('0,0'));
    expect(mockOnClick).toHaveBeenCalled();
  });
  it('Context menu handler on a cell works fine', () => {
    render(<GameWithHooks />);
    userEvent.click(screen.getByTestId('0,0'), { button: 2 });
    expect(mockOnContextMenu).toHaveBeenCalled();
  });
  it('Reset handler works fine', () => {
    render(<GameWithHooks />);
    userEvent.click(screen.getByRole('button'));
    expect(mockOnReset).toHaveBeenCalled();
  });
  it('Change level works fine', () => {
    render(<GameWithHooks />);
    userEvent.selectOptions(screen.getByRole('combobox'), 'intermediate');
    expect(mockHistoryPush).toHaveBeenCalledWith({
      search: `?${new URLSearchParams({ level: 'intermediate' }).toString()}`,
    });
  });
  it('Level in search params works fine', () => {
    (useQuery as jest.Mock).mockReturnValue({ get: () => 'intermediate' });
    render(<GameWithHooks />);
    const intermediateOption = screen.queryByText('intermediate');
    expect(intermediateOption).toBeInTheDocument();
  });
  it('Game over reset the game state', () => {
    render(<GameWithHooks />);
    userEvent.click(screen.getByText('🙁'));
    expect(mockOnReset).toHaveBeenCalled();
  });
});
