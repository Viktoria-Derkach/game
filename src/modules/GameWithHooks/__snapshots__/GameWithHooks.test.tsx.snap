// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameWithHooks test cases Render game field by default 1`] = `
<DocumentFragment>
  <div
    class="css-1gexv9m"
  >
    <div
      class="css-d2q1pa"
    >
      undefined
    </div>
    <select
      class="css-55qvia"
    >
      <option
        class="css-zcmdxr"
        value="beginner"
      >
        beginner
      </option>
      <option
        class="css-zcmdxr"
        value="intermediate"
      >
        intermediate
      </option>
      <option
        class="css-zcmdxr"
        value="expert"
      >
        expert
      </option>
    </select>
    <button
      class="css-1bva717"
    >
      🙂
    </button>
    <div
      class="css-d2q1pa"
    >
      NaN
    </div>
  </div>
  <div
    class="css-1p4hvro"
  >
    🙁
  </div>
  <div
    class="css-da0epy"
    role="grid"
    size="2"
  >
    <div
      class="css-1tt7j02"
      data-testid="0,0"
      role="cell"
    >
      10
    </div>
    <div
      class="css-1tt7j02"
      data-testid="0,1"
      role="cell"
    >
      10
    </div>
    <div
      class="css-1tt7j02"
      data-testid="1,0"
      role="cell"
    >
      10
    </div>
    <div
      class="css-1tt7j02"
      data-testid="1,1"
      role="cell"
    >
      10
    </div>
  </div>
</DocumentFragment>
`;
